src/fiber/fiber.mli

```diff
@@ -228,20 +227,68 @@ end
 with type 'a fiber := 'a t
 
 module Mutex : sig
-  type 'a fiber = 'a t
+  type 'a fiber := 'a t
 
   type t
 
   val create : unit -> t
 
   val with_lock : t -> (unit -> 'a fiber) -> 'a fiber
 end
-with type 'a fiber := 'a t
+
+(** Dealing with cancellation *)
+module Cancel : sig
+  (** Fiber cancellation provides a way to cancel a running fiber. The
+      cancellation mechanism is not builtin the core of fiber and it is not
+      possible to cancel an arbitrary fiber.
+
+      Instead, cancellation is fully explicit. A careful long running fiber may
+      explicitely support cancellation by checking at strategic points if it has
+      been cancelled or not and reacting accordingly. *)
+
+  type 'a fiber := 'a t
+
+  (** A cancellation token. *)
+  type t
+
+  val create : unit -> t
+
+  (** Mark the given cancellation token as cancelled. *)
+  val cancel : t -> unit fiber
+
+  (** [set t f] sets the current cancellation token to [t] while running [f ()]. *)
+  val set : t -> (unit -> 'a fiber) -> 'a fiber
+
+  (** Return whether the current cancellation token has been cancelled. Return
+      [false] if the current fiber doesn't have a cancellation token. *)
+  val cancelled : bool fiber
+
+  type 'a cancelled_or_not =
+    | Cancelled of 'a
+    | Not_cancelled
+
+  (** [with_cancel_handler ~on_cancel f] runs [f ()] with a cancellation
+      handler. If the current cancellation token is cancelled during the
+      execution of [f], then [on_cancel] is called.
+
+      The aim of [on_cancel] is to somehow cut short the execution of [f]. A
+      typical example is a function running an external command. [on_cancel]
+      might send a [KILL] signal to the command to abort its execution.
+
+      If [f ()] finished before the cancellation token is cancelled, then
+      [on_cancel] will never be invoked.
+
+      If the current fiber has no cancellation token, this just executes [f ()]. *)
+  val with_cancel_handler :
+       (unit -> 'a fiber)
+    -> on_cancel:(unit -> 'b fiber)
+    -> ('a * 'b cancelled_or_not) fiber
+end
 
 module Throttle : sig
   (** Limit the number of jobs *)
 
-  type 'a fiber = 'a t
+  type 'a fiber := 'a t
```

Comment:
Which version of OCaml was this feature introduced? We might need to bump the minimum version.

Reply:
Ah, that's annoying. I'll check the CI.